File: src/main/java/me/testaccount666/serversystem/commands/executables/ping/CommandPing.java
package me.testaccount666.serversystem.commands.executables.ping;

import me.testaccount666.serversystem.commands.ServerSystemCommand;
import me.testaccount666.serversystem.commands.executables.AbstractServerSystemCommand;
import me.testaccount666.serversystem.managers.PermissionManager;
import me.testaccount666.serversystem.userdata.User;
import org.bukkit.command.Command;
import org.bukkit.entity.Player;

import static me.testaccount666.serversystem.utils.MessageBuilder.command;
import static me.testaccount666.serversystem.utils.MessageBuilder.general;

@ServerSystemCommand(name = "ping")
public class CommandPing extends AbstractServerSystemCommand {

    @Override
    public void execute(User commandSender, Command command, String label, String... arguments) {
        if (!checkBasePermission(commandSender, "Ping.Use")) return;
        if (handleConsoleWithNoTarget(commandSender, arguments)) return;

        var targetUserOptional = getTargetUser(commandSender, arguments);
        if (targetUserOptional.isEmpty()) {
            general("PlayerNotFound", commandSender).target(arguments[0]).build();
            return;
        }
        var targetUser = targetUserOptional.get();
        var targetPlayer = targetUser.getPlayer();
        var isSelf = targetUser == commandSender;

        if (!isSelf && !checkOtherPermission(commandSender, "Ping.Other", targetPlayer.getName())) return;

        var messagePath = isSelf? "Ping.Success" : "Ping.SuccessOther";
        command(messagePath, commandSender).target(targetPlayer.getName())
                .postModifier(message -> message.replace("<PING>", "${targetPlayer.getPing()}ms")).build();
    }

    @Override
    public boolean hasCommandAccess(Player player, Command command) {
        return PermissionManager.hasCommandPermission(player, "Ping.Use", false);
    }
}

File: src/main/kotlin/me/testaccount666/serversystem/commands/executables/ping/CommandPing.kt
package me.testaccount666.serversystem.commands.executables.ping

import me.testaccount666.serversystem.commands.ServerSystemCommand
import me.testaccount666.serversystem.commands.executables.AbstractServerSystemCommand
import me.testaccount666.serversystem.extensions.commandMsg
import me.testaccount666.serversystem.extensions.generalMsg
import me.testaccount666.serversystem.userdata.User
import org.bukkit.command.Command

@ServerSystemCommand("ping")
class CommandPing : AbstractServerSystemCommand() {
    override fun getUsagePermission(command: Command) = "Ping.Use"
    override fun getSyntaxPath(command: Command?) = "Ping"

    override fun execute(commandSender: User, command: Command, label: String, vararg arguments: String) {
        if (isConsoleWithNoTarget(commandSender, getSyntaxPath(command), label, arguments = arguments)) return

        val targetUser = getTargetUser(commandSender, arguments = arguments) ?: run {
            commandSender.generalMsg("PlayerNotFound") { target(arguments[0]) }
            return
        }

        val targetPlayer = targetUser.getPlayer()!!
        val isSelf = targetUser === commandSender

        if (!isSelf && !checkPermission(commandSender, "Ping.Other", targetPlayer.name)) return

        val messagePath = if (isSelf) "Ping.Success" else "Ping.Other"
        commandSender.commandMsg(messagePath) {
            target(targetPlayer.name)
            postModifier { it.replace("<PING>", "${targetPlayer.ping}ms") }
        }
    }
}
